File: src/main/java/com/handwoong/everyonewaiter/service/MemberServiceImpl.java
package com.handwoong.everyonewaiter.service;

import static com.handwoong.everyonewaiter.enums.ErrorCode.MEMBER_NOT_FOUND;
import static com.handwoong.everyonewaiter.enums.ErrorCode.MEMBER_OR_PHONE_EXISTS;
import static com.handwoong.everyonewaiter.enums.ErrorCode.NOT_MATCH_PASSWORD;

import com.handwoong.everyonewaiter.domain.Member;
import com.handwoong.everyonewaiter.dto.MemberDto;
import com.handwoong.everyonewaiter.dto.OnlyMsgResDto;
import com.handwoong.everyonewaiter.exception.CustomException;
import com.handwoong.everyonewaiter.repository.MemberRepository;
import java.util.List;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Slf4j
@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class MemberServiceImpl implements MemberService {

    private final MemberRepository memberRepository;

    private final PasswordEncoder passwordEncoder;

    @Override
    @Transactional
    public Long register(MemberDto.RequestDto memberDto) {
        String username = memberDto.getUsername();
        String phoneNumber = memberDto.getPhoneNumber();

        isExistsUsernameOrPhone(username, phoneNumber);

        Member member = Member.createMember(memberDto);
        member.encodePassword(passwordEncoder.encode(member.getPassword()));
        memberRepository.save(member);

        return member.getId();
    }

    @Override
    public MemberDto.ResponseDto findMember(Long memberId) {
        Member member = findById(memberId);

        return MemberDto.ResponseDto.from(member);
    }

    @Override
    public MemberDto.ResponseDto findMemberByUsername(String username) {
        Member member = findByUsername(username);

        return MemberDto.ResponseDto.from(member);
    }

    @Override
    public List<MemberDto.ResponseDto> findMemberList() {
        List<Member> members = memberRepository.findAll();

        return members.stream()
                .map(MemberDto.ResponseDto::from)
                .toList();
    }

    @Override
    @Transactional
    public OnlyMsgResDto changePassword(String username, MemberDto.PwdRequestDto passwordDto) {
        Member member = findByUsername(username);
        matchPassword(member, passwordDto.getCurrentPassword());
        member.encodePassword(passwordEncoder.encode(passwordDto.getNewPassword()));

        return new OnlyMsgResDto("success");
    }

    @Override
    @Transactional
    public OnlyMsgResDto deleteMember(String username, MemberDto.PwdRequestDto passwordDto) {
        Member member = findByUsername(username);
        matchPassword(member, passwordDto.getCurrentPassword());
        memberRepository.deleteByUsername(member.getUsername());

        return new OnlyMsgResDto("success");
    }

    private void isExistsUsernameOrPhone(String username, String phoneNumber) {
        boolean isExists = memberRepository.existsByUsernameOrPhoneNumber(username, phoneNumber);

        if (isExists) {
            throw new CustomException(MEMBER_OR_PHONE_EXISTS);
        }
    }

    private Member findById(Long memberId) {
        return memberRepository.findById(memberId)
                .orElseThrow(() -> new CustomException(MEMBER_NOT_FOUND));
    }

    private Member findByUsername(String username) {
        return memberRepository.findByUsername(username)
                .orElseThrow(() -> new CustomException(MEMBER_NOT_FOUND));
    }

    private void matchPassword(Member member, String password) {
        boolean isCorrectPassword = passwordEncoder.matches(password, member.getPassword());

        if (!isCorrectPassword) {
            throw new CustomException(NOT_MATCH_PASSWORD);
        }
    }
}

File: src/main/kotlin/com/handwoong/everyonewaiter/service/member/MemberServiceImpl.kt
package com.handwoong.everyonewaiter.service.member

import com.handwoong.everyonewaiter.config.security.JwtTokenProvider
import com.handwoong.everyonewaiter.domain.member.Member
import com.handwoong.everyonewaiter.dto.member.*
import com.handwoong.everyonewaiter.exception.ErrorCode.*
import com.handwoong.everyonewaiter.repository.member.MemberRepository
import com.handwoong.everyonewaiter.util.findByIdOrThrow
import com.handwoong.everyonewaiter.util.throwFail
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class MemberServiceImpl(
    private val memberRepository: MemberRepository,
    private val authenticationManagerBuilder: AuthenticationManagerBuilder,
    private val jwtTokenProvider: JwtTokenProvider,
    private val passwordEncoder: PasswordEncoder,
) : MemberService {

    override fun login(memberDto: MemberLoginRequest): TokenResponse {
        val authenticationToken = UsernamePasswordAuthenticationToken(memberDto.username, memberDto.password)
        val authenticate = authenticationManagerBuilder.`object`.authenticate(authenticationToken)
        return jwtTokenProvider.generateToken(authenticate)
    }

    @Transactional
    override fun register(memberDto: MemberCreateRequest) {
        existsUsernameOrPhoneNumber(
            username = memberDto.username,
            phoneNumber = memberDto.phoneNumber,
        )

        val createMember = Member.createMember(memberDto)
        createMember.encodePassword(passwordEncoder.encode(createMember.password))
        memberRepository.save(createMember)
    }

    override fun findMemberById(memberId: Long): MemberResponse {
        val findMember = memberRepository.findByIdOrThrow(memberId)
        return MemberResponse.of(findMember)
    }

    override fun findMemberByUsername(username: String): MemberResponse {
        val findMember = findByUsername(username)
        return MemberResponse.of(findMember)
    }

    override fun findAllMemberList(): List<MemberResponse> {
        val memberList = memberRepository.findAll()
        return memberList.map { member -> MemberResponse.of(member) }
    }

    @Transactional
    override fun changePassword(username: String, passwordDto: PasswordRequest) {
        val findMember = findByUsername(username)
        matchPassword(passwordDto.currentPassword, findMember.password)
        findMember.encodePassword(passwordEncoder.encode(passwordDto.newPassword))
    }

    @Transactional
    override fun deleteMember(username: String, passwordDto: PasswordRequest) {
        val findMember = findByUsername(username)
        matchPassword(passwordDto.currentPassword, findMember.password)
        memberRepository.delete(findMember)
    }

    private fun existsUsernameOrPhoneNumber(username: String, phoneNumber: String? = null) {
        if (memberRepository.existsMember(username, phoneNumber)) {
            throwFail(USERNAME_OR_PHONE_EXISTS)
        }
    }

    private fun findByUsername(username: String): Member {
        return memberRepository.findMember(username) ?: throwFail(MEMBER_NOT_FOUND)
    }

    private fun matchPassword(userInputPassword: String, currentPassword: String) {
        if (!passwordEncoder.matches(userInputPassword, currentPassword)) {
            throwFail(NOT_MATCH_PASSWORD)
        }
    }

}
